packages/framework/src/Model/Mail/MailTemplateVariables.php

```diff
@@ -0,0 +1,231 @@
+
+class MailTemplateVariables
+{
+    /**
+     * Variable will be replaced in email body
+     */
+    public const CONTEXT_BODY = 1;
+    /**
+     * Variable will be replaced in email subject
+     */
+    public const CONTEXT_SUBJECT = 2;
+    /**
+     * Variable will be replaced in email body and subject
+     */
+    public const CONTEXT_BOTH = 3;
+    /**
+     * Variable is not required
+     */
+    public const REQUIRED_NOWHERE = 0;
+    /**
+     * Variable is required in email body
+     */
+    public const REQUIRED_BODY = 1;
+    /**
+     * Variable is required in email subject
+     */
+    public const REQUIRED_SUBJECT = 2;
+    /**
+     * Variable is required in email body and subject
+     */
+    public const REQUIRED_BOTH = 3;
+
+    /**
+     * @var string
+     */
+    protected $readableName;
+
+    /**
+     * @var array
+     */
+    protected $variables = [];
+
+    /**
+     * @var string[]
+     */
+    protected $bodyVariables = [];
+
+    /**
+     * @var string[]
+     */
+    protected $subjectVariables = [];
+
+    /**
+     * @var string[]
+     */
+    protected $requiredBodyVariables = [];
+
+    /**
+     * @var string[]
+     */
+    protected $requiredSubjectVariables = [];
+
+    /**
+     * @var string|null
+     */
+    protected $type;
+
+    /**
+     * @param string $readableName
+     * @param string|null $type
+     */
+    public function __construct(string $readableName, ?string $type = null)
+    {
+        $this->readableName = $readableName;
+        $this->type = $type;
+    }
+
+    /**
+     * @param string $readableName
+     * @return \Shopsys\FrameworkBundle\Model\Mail\MailTemplateVariables
+     */
+    public function withNewName(string $readableName): self
+    {
+        $clone = clone $this;
+
+        $clone->readableName = $readableName;
+
+        return $clone;
+    }
+
+    /**
+     * @param string $variable
+     * @param string $label
+     * @param int $context one of CONTEXT_* constants
+     * @param int $required one of REQUIRED_* constants
+     * @return $this
+     */
+    public function addVariable(string $variable, string $label, int $context = self::CONTEXT_BOTH, int $required = self::REQUIRED_NOWHERE): self
+    {
+        if (array_key_exists($variable, $this->variables)) {
+            throw new InvalidMailTemplateVariablesConfigurationException(sprintf('Variable "%s" is already registered.', $variable));
+        }
+
+        $this->variables[$variable] = $label;
+        $this->addVariableToSections($variable, $context);
+        $this->addVariableToRequiredSections($variable, $context, $required);
+
+        return $this;
+    }
+
+    /**
+     * @param string $variable
+     * @param int $context
+     */
+    protected function addVariableToSections(string $variable, int $context): void
+    {
+        switch ($context) {
+            case self::CONTEXT_BOTH:
```

Comment:
I wonder if we shouldn't use a static instead of a self. What do you think?

Reply:
static cannot be used, because the constant is public. 
when anyone changes the value of the constant in extended class (not gonna happen here probably), here would be the new value. 
But anywhere else where `MailTemplateVariables::CONTEXT_BOTH` is used, the value would be still the same and it can lead to errors.